ai_ta_backend/utils/pubmed_extraction.py

```diff
@@ -19,808 +19,730 @@
-    # os.remove(gz_filepath)
-    print("Finished file: ", file)
-
-  except Exception as e:
-    print("Error processing XML file: ", e)
-    with open("errors.txt", 'a') as f:
-      f.write("Error in proccesing XML file: " + file + str(e) + "\n")
-
-  end_time = time.monotonic()
-
-  POSTHOG.capture(distinct_id="pubmed_extraction",
-                  event="processPubmedXML",
-                  properties={
-                      "total_runtime": end_time - start_time,
-                  })
-
-
-def downloadXML(ftp_address: str, ftp_path: str, file: str, local_dir: str):
-  """
-    Downloads a .gz XML file from the FTP baseline folder and stores it in the local directory.
-    Args:
-        ftp_address: FTP server address.
-        ftp_path: Path to the FTP folder.
-        file: File to download.
-        local_dir: Local directory to store the downloaded file.
-    Returns:
-        local_filepath: Path to the downloaded file.
+      print("Error extracting XML file: ", e)
+      return None
+
+  def extractMetadataFromXML(self, xml_filepath: str, dir: str, error_file: str):
     """
-  try:
-    # create local directory if it doesn't exist
-    os.makedirs(local_dir, exist_ok=True)
-
-    # connect to the FTP server
-    ftp = ftplib.FTP(ftp_address)
-    ftp.login()
-    ftp.cwd(ftp_path)
-
-    local_filepath = os.path.join(local_dir, file)
-    with open(local_filepath, 'wb') as f:
-      ftp.retrbinary('RETR ' + file, f.write)
-
-    # print(f"Downloaded {file} to {local_filepath}")
-
-    ftp.quit()
-    return local_filepath
-  except Exception as e:
-    print("Error downloading file: ", e)
-    return None
-
-
-def getFileList(ftp_address: str, ftp_path: str, extension: str = ".gz"):
-  """
-    Returns a list of .gz files in the FTP baseline folder.
-    Args:
-        ftp_address: FTP server address.
-        ftp_path: Path to the FTP folder.
-        extension: File extension to filter for.
-    Returns:
-        gz_files: List of .gz files in the FTP folder.
+        Extracts article details from the XML file and stores it in a dictionary.
+        Details extracted: PMID, PMCID, DOI, ISSN, journal title, article title,
+        last revised date, published date, abstract.
+        Args:
+            xml_filepath: Path to the XML file.
+        Returns:
+            metadata: List of dictionaries containing metadata for each article.
+        """
+    print("inside extractMetadataFromXML()")
+    start_time = time.monotonic()
+    try:
+      # create a directory to store papers
+      os.makedirs(dir, exist_ok=True)
+
+      tree = ET.parse(xml_filepath)
+      root = tree.getroot()
+      metadata = []
+
+      with concurrent.futures.ProcessPoolExecutor() as executor:
+        futures = []
+        article_items = list(item for item in root.iter('PubmedArticle'))  # Convert generator to list
+
+        for item in article_items:
+          future = executor.submit(self.processArticleItem, item, dir, error_file)
```

Comment:
The class structure of the script is leading to pickle errors like below - 
E.g.: `cannot pickle '_thread.lock' object`